tests/authenticate.go

```diff
@@ -0,0 +1,428 @@
+		},
+	}
+	client := ctx.Client(ctx.HttpClient(transport))
+
+	resp, err := client.R().
+		SetHeader("Content-Type", "application/json").
+		Post("/authenticate?method=cert")
+
+	if err != nil {
+		return nil, errors.Errorf("failed to authenticate via CERT: %s", err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return nil, errors.Errorf("failed to authenticate via CERT: invalid response code encountered, got %d, expected %d", resp.StatusCode(), http.StatusOK)
+	}
+
+	if err = sess.parseSessionInfoFromResponse(resp); err != nil {
+		return nil, err
+	}
+
+	return sess, nil
+}
+
+type updbAuthenticator struct {
+	Username string
+	Password string
+}
+
+func (authenticator *updbAuthenticator) Authenticate(ctx *TestContext) (*session, error) {
+	sess := &session{
+		authenticator: authenticator,
+		testContext:   ctx,
+	}
+
+	sess.authenticatedRequests = authenticatedRequests{
+		testContext: ctx,
+		session:     sess,
+	}
+
+	body := gabs.New()
+	_, _ = body.SetP(authenticator.Username, "username")
+	_, _ = body.SetP(authenticator.Password, "password")
+
+	resp, err := ctx.DefaultClient().R().
+		SetHeader("Content-Type", "application/json").
+		SetBody(body.String()).
+		Post("/authenticate?method=password")
+
+	if err != nil {
+		return nil, errors.Errorf("failed to authenticate via UPDB as %v: %s", authenticator, err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return nil, errors.Errorf("failed to authenticate via UPDB as %v: invalid response code encountered, got %d, expected %d", authenticator, resp.StatusCode(), http.StatusOK)
+	}
+
+	if err = sess.parseSessionInfoFromResponse(resp); err != nil {
+		return nil, err
+	}
+
+	return sess, nil
+}
+
+type session struct {
+	authenticator authenticator
+	token         string
+	identityId    string
+	testContext   *TestContext
+	authenticatedRequests
+}
+
+func (sess *session) newRequest(ctx *TestContext) *resty.Request {
+	req := ctx.newRequest()
+	if sess.token != "" {
+		req.SetHeader(constants.ZitiSession, sess.token)
+	}
+	return req
+}
+
+func (sess *session) parseSessionInfoFromResponse(response *resty.Response) error {
+	respBody := response.Body()
+
+	if len(respBody) == 0 {
+		return errors.Errorf("failed to authenticate via UPDB %v: encountered zero length body", sess.authenticator)
+	}
+
+	bodyContainer, err := gabs.ParseJSON(respBody)
+
+	if err != nil {
+		return errors.Errorf("failed to authenticate via UPDB as %v: failed to parse response: %s", sess.authenticator, err)
+	}
+
+	if sessionToken, ok := bodyContainer.Path("data.token").Data().(string); ok {
+		sess.token = sessionToken
+	} else {
+		return errors.Errorf("failed to authenticate via UPDB as %v: failed to find session token", sess.authenticator)
+	}
+
+	if identityId, ok := bodyContainer.Path("data.identity.id").Data().(string); ok {
+		sess.identityId = identityId
+	} else {
+		return errors.Errorf("failed to authenticate via UPDB as %v: failed to find identity id", sess.authenticator)
```

Comment:
i tend to not like providing any kind of information as to why authentication fails. can we just say "could not log in"? or something?

Reply:
oh hey that's in a test. nevermind :)